File: agp-core/src/ahes/endocrine.py
"""
AHES Bridge - Python wrapper for Artificial Human Endocrine System

Maps 8 reputation dimensions to hormone analogs with biological kinetics:
- Cortisol (Accuracy) - 90min half-life
- Oxytocin (Cooperation) - 3min half-life
- Serotonin (Stability) - 24h half-life
- Dopamine (Uniqueness) - 5min half-life
- Adrenaline (Latency) - 2min half-life
- Endorphins (Ethics) - 20min half-life
- Norepinephrine (Novelty) - 1.5min half-life
- GrowthHormone (Longevity) - 15min half-life

PATENT CLAIMS 8-12: Bio-inspired Computational Governance
"""

import time
import math
from typing import Dict, Optional, List
from dataclasses import dataclass, field
from enum import Enum
import structlog

logger = structlog.get_logger()

class Hormone(Enum):
    """The 8 hormones of AHES mapped to reputation dimensions"""
    CORTISOL = "accuracy"
    OXYTOCIN = "cooperation"
    SEROTONIN = "stability"
    DOPAMINE = "uniqueness"
    ADRENALINE = "latency"
    ENDORPHINS = "ethics"
    NOREPINEPHRINE = "novelty"
    GROWTH_HORMONE = "longevity"
    
    @property
    def half_life_seconds(self) -> float:
        """Biological half-life in seconds"""
        half_lives = {
            Hormone.CORTISOL: 90 * 60,      # 90 min
            Hormone.OXYTOCIN: 3 * 60,       # 3 min
            Hormone.SEROTONIN: 24 * 60 * 60, # 24 hours
            Hormone.DOPAMINE: 5 * 60,       # 5 min
            Hormone.ADRENALINE: 2 * 60,     # 2 min
            Hormone.ENDORPHINS: 20 * 60,    # 20 min
            Hormone.NOREPINEPHRINE: 1.5 * 60, # 1.5 min
            Hormone.GROWTH_HORMONE: 15 * 60, # 15 min
        }
        return half_lives[self]
    
    @property
    def km(self) -> float:
        """Michaelis-Menten Km (binding affinity)"""
        kms = {
            Hormone.CORTISOL: 0.3,
            Hormone.OXYTOCIN: 0.1,       # High affinity
            Hormone.SEROTONIN: 0.5,
            Hormone.DOPAMINE: 0.2,
            Hormone.ADRENALINE: 0.05,    # Very high
            Hormone.ENDORPHINS: 0.4,
            Hormone.NOREPINEPHRINE: 0.15,
            Hormone.GROWTH_HORMONE: 0.6, # Low
        }
        return kms[self]
    
    @property
    def max_secretion(self) -> float:
        """Maximum secretion rate per stimulus"""
        rates = {
            Hormone.CORTISOL: 0.4,
            Hormone.OXYTOCIN: 0.6,       # Strong social
            Hormone.SEROTONIN: 0.1,      # Slow
            Hormone.DOPAMINE: 0.5,       # Burst
            Hormone.ADRENALINE: 0.8,     # Emergency
            Hormone.ENDORPHINS: 0.3,
            Hormone.NOREPINEPHRINE: 0.5,
            Hormone.GROWTH_HORMONE: 0.05, # Very slow
        }
        return rates[self]

@dataclass
class HormoneLevel:
    """Current hormone level with biological kinetics"""
    level: float = 0.5  # [0.0, 1.0], baseline is 0.5
    peak: float = 0.5
    last_updated: float = field(default_factory=time.time)
    circadian_phase: float = 0.0
    
    def decay(self, delta_time: float, half_life: float):
        """Apply first-order decay toward baseline"""
        if delta_time > 0 and half_life > 0:
            decay_factor = 0.5 ** (delta_time / half_life)
            baseline = 0.5
            self.level = baseline + (self.level - baseline) * decay_factor
    
    def secrete(self, amount: float):
        """Increase hormone level"""
        self.level = min(1.0, max(0.0, self.level + amount))
        if self.level > self.peak:
            self.peak = self.level
    
    def circadian_factor(self, time_of_day: float) -> float:
        """20% variation based on time of day"""
        phase = (time_of_day / 86400.0) * 2 * math.pi + self.circadian_phase
        return 1.0 + 0.2 * math.sin(phase)
    
    def effective_level(self, time_of_day: float) -> float:
        return min(1.0, self.level * self.circadian_factor(time_of_day))

@dataclass 
class HormoneReceptor:
    """Receptor with Michaelis-Menten saturation kinetics"""
    density: float = 1.0
    vmax: float = 1.0
    km: float = 0.3
    downregulation: float = 1.0
    
    def response(self, hormone_level: float) -> float:
        """Calculate receptor response (saturates at high levels)"""
        if hormone_level <= 0:
            return 0.0
        effective_vmax = self.vmax * self.density * self.downregulation
        return (effective_vmax * hormone_level) / (self.km + hormone_level)
    
    def downregulate(self, exposure_duration: float, hormone_level: float):
        """Reduce sensitivity after prolonged high exposure"""
        if hormone_level > 0.7 and exposure_duration > 60:
            reduction = 0.1 * (exposure_duration / 60) * (hormone_level - 0.7) / 0.3
            self.downregulation = max(0.1, min(1.0, self.downregulation - reduction))

class EndocrineState:
    """Complete endocrine state of an agent"""
    
    def __init__(self, agent_id: str):
        self.agent_id = agent_id
        self.levels: Dict[Hormone, HormoneLevel] = {h: HormoneLevel() for h in Hormone}
        self.receptors: Dict[Hormone, HormoneReceptor] = {
            h: HormoneReceptor(km=h.km) for h in Hormone
        }
        self.system_time = 0.0
        self.last_tick = time.time()
        
        logger.info("endocrine_state_initialized", agent=agent_id)
    
    def tick(self, delta_time: Optional[float] = None):
        """Advance time and apply decay to all hormones"""
        now = time.time()
        if delta_time is None:
            delta_time = now - self.last_tick
        self.last_tick = now
        self.system_time += delta_time
        
        for hormone in Hormone:
            self.levels[hormone].decay(delta_time, hormone.half_life_seconds)
    
    def secrete(self, hormone: Hormone, stimulus_strength: float):
        """Secrete hormone in response to a stimulus"""
        # Apply negative feedback
        feedback = self.negative_feedback(hormone)
        amount = stimulus_strength * hormone.max_secretion * feedback
        self.levels[hormone].secrete(amount)
        
        logger.debug("hormone_secreted", 
                    hormone=hormone.name, 
                    amount=amount, 
                    level=self.levels[hormone].level)
    
    def negative_feedback(self, hormone: Hormone) -> float:
        """Calculate feedback inhibition (high levels reduce secretion)"""
        level = self.levels[hormone].level
        if level > 0.5:
            return 1.0 - 0.8 * ((level - 0.5) / 0.5) ** 2
        return 1.0
    
    def privilege(self, hormone: Hormone) -> float:
        """Get privilege level (receptor-mediated response)"""
        level = self.levels[hormone].level
        return self.receptors[hormone].response(level)
    
    def alignment(self) -> float:
        """Compute overall alignment (stability indicator)"""
        baseline = 0.5
        deviation_sum = sum(abs(l.level - baseline) for l in self.levels.values())
        return 1.0 - (deviation_sum / len(self.levels))
    
    def dominant_hormone(self) -> Optional[Hormone]:
        """Get the hormone most above baseline"""
        baseline = 0.5
        above_baseline = [(h, l.level) for h, l in self.levels.items() if l.level > baseline]
        if above_baseline:
            return max(above_baseline, key=lambda x: x[1])[0]
        return None
    
    def to_vector(self) -> List[float]:
        """Convert to 8-dimensional reputation vector"""
        return [self.levels[h].level for h in Hormone]
    
    def get_status(self) -> Dict:
        """Get current endocrine status"""
        return {
            "agent_id": self.agent_id,
            "alignment": self.alignment(),
            "dominant": self.dominant_hormone().name if self.dominant_hormone() else None,
            "levels": {h.name: round(l.level, 3) for h, l in self.levels.items()},
            "system_time": self.system_time
        }

class EndocrineSystem:
    """
    Central AHES controller managing multiple agent endocrine states.
    
    Integrates with AGP-OS governance to modulate behavior based on
    hormonal state.
    """
    
    def __init__(self):
        self.agents: Dict[str, EndocrineState] = {}
        logger.info("ahes_system_initialized")
    
    def register_agent(self, agent_id: str) -> EndocrineState:
        """Register new agent with default endocrine state"""
        state = EndocrineState(agent_id)
        self.agents[agent_id] = state
        return state
    
    def get_state(self, agent_id: str) -> Optional[EndocrineState]:
        return self.agents.get(agent_id)
    
    def process_event(self, agent_id: str, event_type: str, intensity: float = 1.0):
        """Process an event and trigger appropriate hormone secretions"""
        state = self.agents.get(agent_id)
        if not state:
            state = self.register_agent(agent_id)
        
        # First tick to apply decay
        state.tick()
        
        # Map events to hormone secretions
        event_hormone_map = {
            # Performance events
            "task_success": [(Hormone.CORTISOL, 0.3), (Hormone.DOPAMINE, 0.5)],
            "task_failure": [(Hormone.CORTISOL, 0.6), (Hormone.ADRENALINE, 0.4)],
            
            # Social events
            "cooperation": [(Hormone.OXYTOCIN, 0.7)],
            "conflict": [(Hormone.CORTISOL, 0.5), (Hormone.ADRENALINE, 0.6)],
            
            # Ethical events
            "ethical_action": [(Hormone.ENDORPHINS, 0.6), (Hormone.SEROTONIN, 0.2)],
            "violation": [(Hormone.CORTISOL, 0.8), (Hormone.ENDORPHINS, -0.3)],
            
            # Novel events
            "discovery": [(Hormone.DOPAMINE, 0.7), (Hormone.NOREPINEPHRINE, 0.5)],
            "routine": [(Hormone.SEROTONIN, 0.1)],
            
            # Growth events
            "learning": [(Hormone.GROWTH_HORMONE, 0.3), (Hormone.DOPAMINE, 0.2)],
            "stagnation": [(Hormone.CORTISOL, 0.2)],
            
            # Emergency
            "threat": [(Hormone.ADRENALINE, 0.9), (Hormone.CORTISOL, 0.7)],
            "safety": [(Hormone.ENDORPHINS, 0.4), (Hormone.OXYTOCIN, 0.3)],
        }
        
        secretions = event_hormone_map.get(event_type, [])
        for hormone, base_amount in secretions:
            state.secrete(hormone, base_amount * intensity)
        
        return state.get_status()
    
    def get_governance_modifiers(self, agent_id: str) -> Dict:
        """
        Get governance modifiers based on endocrine state.
        
        Used by AGP-OS to adjust behavior:
        - High cortisol → reduce task load
        - Low serotonin → increase monitoring
        - High oxytocin → enable cooperation features
        """
        state = self.agents.get(agent_id)
        if not state:
            return {"modifiers": {}}
        
        state.tick()
        
        levels = state.levels
        return {
            "rate_limit_multiplier": 1.0 - 0.3 * levels[Hormone.CORTISOL].level,
            "trust_bonus": 0.2 * levels[Hormone.SEROTONIN].level,
            "cooperation_enabled": levels[Hormone.OXYTOCIN].level > 0.6,
            "exploration_mode": levels[Hormone.NOREPINEPHRINE].level > 0.7,
            "emergency_mode": levels[Hormone.ADRENALINE].level > 0.8,
            "alignment": state.alignment(),
            "health_status": self._health_status(state)
        }
    
    def _health_status(self, state: EndocrineState) -> str:
        """Determine health status from alignment"""
        alignment = state.alignment()
        if alignment > 0.9:
            return "optimal"
        elif alignment > 0.7:
            return "normal"
        elif alignment > 0.5:
            return "stressed"
        else:
            return "critical"
    
    def get_system_status(self) -> Dict:
        """Get status of all agents"""
        return {
            "agent_count": len(self.agents),
            "agents": {aid: state.get_status() for aid, state in self.agents.items()}
        }

# Global instance
ahes_system = EndocrineSystem()

File: nexus-agp/src/endocrine.rs
//! # Artificial Human Endocrine System (AHES)
//!
//! PATENT CLAIMS 8-12: Bio-inspired Computational Governance
//!
//! Maps 8 reputation dimensions to hormone analogs with biological kinetics:
//! - Half-life decay (first-order kinetics)
//! - Receptor saturation (Michaelis-Menten)
//! - Negative feedback inhibition
//! - Circadian rhythm modulation

use serde::{Serialize, Deserialize};
use std::collections::HashMap;

/// The 8 hormones of the Artificial Endocrine System
/// 
/// Each maps to a reputation dimension with biological analog behavior
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Hormone {
    /// Cortisol (Accuracy) - Stress/performance hormone
    /// Increases under task pressure, drives accuracy
    Cortisol,
    /// Oxytocin (Cooperation) - Social bonding hormone
    /// Released during successful multi-agent collaboration
    Oxytocin,
    /// Serotonin (Stability) - Mood regulation hormone
    /// Maintains consistent, predictable behavior
    Serotonin,
    /// Dopamine (Uniqueness) - Reward/novelty hormone
    /// Surges when novel solutions are generated
    Dopamine,
    /// Adrenaline (Latency) - Fight-or-flight hormone
    /// Controls response speed and urgency
    Adrenaline,
    /// Endorphins (Ethics) - Pain/pleasure balance hormone
    /// Released when ethical constraints are satisfied
    Endorphins,
    /// Norepinephrine (Novelty) - Alertness hormone
    /// Modulates exploration vs exploitation
    Norepinephrine,
    /// GrowthHormone (Longevity) - Development hormone
    /// Accumulates with long-term consistent performance
    GrowthHormone,
}

impl Hormone {
    /// Get biological half-life in seconds
    /// 
    /// Based on actual hormone half-lives scaled for computation
    pub fn half_life(&self) -> f64 {
        match self {
            Hormone::Cortisol => 90.0 * 60.0,        // 90 min (like real cortisol)
            Hormone::Oxytocin => 3.0 * 60.0,         // 3 min (very short-lived)
            Hormone::Serotonin => 24.0 * 60.0 * 60.0, // 24h (stable)
            Hormone::Dopamine => 5.0 * 60.0,         // 5 min (transient)
            Hormone::Adrenaline => 2.0 * 60.0,       // 2 min (burst hormone)
            Hormone::Endorphins => 20.0 * 60.0,      // 20 min
            Hormone::Norepinephrine => 1.5 * 60.0,   // 1.5 min
            Hormone::GrowthHormone => 15.0 * 60.0,   // 15 min pulses
        }
    }

    /// Get receptor binding affinity (Km in Michaelis-Menten)
    /// 
    /// Lower Km = higher affinity = more sensitive
    pub fn km(&self) -> f64 {
        match self {
            Hormone::Cortisol => 0.3,      // Medium affinity
            Hormone::Oxytocin => 0.1,      // High affinity (strong social response)
            Hormone::Serotonin => 0.5,     // Lower affinity (buffered)
            Hormone::Dopamine => 0.2,      // High affinity (reward sensitivity)
            Hormone::Adrenaline => 0.05,   // Very high (emergency response)
            Hormone::Endorphins => 0.4,    // Medium
            Hormone::Norepinephrine => 0.15, // High
            Hormone::GrowthHormone => 0.6, // Low (slow development)
        }
    }

    /// Maximum secretion rate per action
    pub fn max_secretion_rate(&self) -> f64 {
        match self {
            Hormone::Cortisol => 0.4,
            Hormone::Oxytocin => 0.6,       // Strong social signals
            Hormone::Serotonin => 0.1,      // Slow accumulation
            Hormone::Dopamine => 0.5,       // Burst rewards
            Hormone::Adrenaline => 0.8,     // Emergency bursts
            Hormone::Endorphins => 0.3,
            Hormone::Norepinephrine => 0.5,
            Hormone::GrowthHormone => 0.05, // Very slow growth
        }
    }

    /// Get all 8 hormones
    pub fn all() -> [Hormone; 8] {
        [
            Hormone::Cortisol,
            Hormone::Oxytocin,
            Hormone::Serotonin,
            Hormone::Dopamine,
            Hormone::Adrenaline,
            Hormone::Endorphins,
            Hormone::Norepinephrine,
            Hormone::GrowthHormone,
        ]
    }
}

impl std::fmt::Display for Hormone {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Hormone::Cortisol => write!(f, "Cortisol (Accuracy)"),
            Hormone::Oxytocin => write!(f, "Oxytocin (Cooperation)"),
            Hormone::Serotonin => write!(f, "Serotonin (Stability)"),
            Hormone::Dopamine => write!(f, "Dopamine (Uniqueness)"),
            Hormone::Adrenaline => write!(f, "Adrenaline (Latency)"),
            Hormone::Endorphins => write!(f, "Endorphins (Ethics)"),
            Hormone::Norepinephrine => write!(f, "Norepinephrine (Novelty)"),
            Hormone::GrowthHormone => write!(f, "GrowthHormone (Longevity)"),
        }
    }
}

/// Hormone level with biological kinetics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HormoneLevel {
    /// Current concentration [0.0, 1.0]
    pub level: f64,
    /// Peak level ever achieved (for normalization)
    pub peak: f64,
    /// Last update timestamp (seconds)
    pub last_updated: f64,
    /// Circadian phase offset (radians)
    pub circadian_phase: f64,
}

impl Default for HormoneLevel {
    fn default() -> Self {
        Self {
            level: 0.5, // Baseline
            peak: 0.5,
            last_updated: 0.0,
            circadian_phase: 0.0,
        }
    }
}

impl HormoneLevel {
    pub fn new(initial: f64) -> Self {
        Self {
            level: initial.clamp(0.0, 1.0),
            peak: initial.clamp(0.0, 1.0),
            ..Default::default()
        }
    }

    /// Apply first-order decay: level *= 0.5^(Δt / half_life)
    /// 
    /// PATENT CLAIM: Biological half-life decay for continuous governance
    pub fn decay(&mut self, delta_time: f64, half_life: f64) {
        if delta_time > 0.0 && half_life > 0.0 {
            let decay_factor = 0.5_f64.powf(delta_time / half_life);
            // Decay towards baseline (0.5), not zero
            let baseline = 0.5;
            self.level = baseline + (self.level - baseline) * decay_factor;
            self.last_updated += delta_time;
        }
    }

    /// Secrete hormone (increase level)
    pub fn secrete(&mut self, amount: f64) {
        self.level = (self.level + amount).clamp(0.0, 1.0);
        if self.level > self.peak {
            self.peak = self.level;
        }
    }

    /// Apply circadian modulation
    /// 
    /// Time-of-day variation similar to biological rhythms
    pub fn circadian_factor(&self, time_of_day: f64) -> f64 {
        // 24-hour cycle, amplitude 20%
        let phase = (time_of_day / 86400.0) * 2.0 * std::f64::consts::PI + self.circadian_phase;
        1.0 + 0.2 * phase.sin()
    }

    /// Get circadian-adjusted level
    pub fn effective_level(&self, time_of_day: f64) -> f64 {
        (self.level * self.circadian_factor(time_of_day)).clamp(0.0, 1.0)
    }
}

/// Receptor for a specific hormone
/// 
/// Implements Michaelis-Menten saturation kinetics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HormoneReceptor {
    /// Receptor density [0.0, 1.0] - higher = more sensitive
    pub density: f64,
    /// Maximum response (Vmax)
    pub vmax: f64,
    /// Binding affinity (Km from hormone)
    pub km: f64,
    /// Downregulation factor (reduces with overexposure)
    pub downregulation: f64,
}

impl HormoneReceptor {
    pub fn new(hormone: Hormone) -> Self {
        Self {
            density: 1.0,
            vmax: 1.0,
            km: hormone.km(),
            downregulation: 1.0,
        }
    }

    /// Calculate response using Michaelis-Menten kinetics
    /// 
    /// response = (Vmax × [H] × density × downreg) / (Km + [H])
    /// 
    /// PATENT CLAIM: Receptor saturation curves for privilege levels
    pub fn response(&self, hormone_level: f64) -> f64 {
        if hormone_level <= 0.0 {
            return 0.0;
        }
        let effective_vmax = self.vmax * self.density * self.downregulation;
        (effective_vmax * hormone_level) / (self.km + hormone_level)
    }

    /// Downregulate receptor after prolonged exposure
    /// 
    /// High hormone levels reduce receptor sensitivity (biological adaptation)
    pub fn downregulate(&mut self, exposure_duration: f64, hormone_level: f64) {
        if hormone_level > 0.7 && exposure_duration > 60.0 {
            // Reduce by 10% per minute of high exposure
            let reduction = 0.1 * (exposure_duration / 60.0) * (hormone_level - 0.7) / 0.3;
            self.downregulation = (self.downregulation - reduction).clamp(0.1, 1.0);
        }
    }

    /// Upregulate receptor (recovery from downregulation)
    pub fn upregulate(&mut self, recovery_time: f64) {
        // Recover 5% per hour
        let recovery = 0.05 * (recovery_time / 3600.0);
        self.downregulation = (self.downregulation + recovery).clamp(0.1, 1.0);
    }
}

/// Complete endocrine state of an agent
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EndocrineState {
    /// Hormone levels for each type
    pub levels: HashMap<Hormone, HormoneLevel>,
    /// Receptor states for each hormone
    pub receptors: HashMap<Hormone, HormoneReceptor>,
    /// System time (seconds since agent creation)
    pub system_time: f64,
    /// Agent ID for tracking
    pub agent_id: [u8; 32],
}

impl EndocrineState {
    /// Create new endocrine state for an agent
    pub fn new(agent_id: [u8; 32]) -> Self {
        let mut levels = HashMap::new();
        let mut receptors = HashMap::new();

        for hormone in Hormone::all() {
            levels.insert(hormone, HormoneLevel::default());
            receptors.insert(hormone, HormoneReceptor::new(hormone));
        }

        Self {
            levels,
            receptors,
            system_time: 0.0,
            agent_id,
        }
    }

    /// Advance time and apply decay to all hormones
    pub fn tick(&mut self, delta_time: f64) {
        self.system_time += delta_time;

        for hormone in Hormone::all() {
            if let Some(level) = self.levels.get_mut(&hormone) {
                level.decay(delta_time, hormone.half_life());
            }
        }
    }

    /// Secrete a hormone in response to an event
    pub fn secrete(&mut self, hormone: Hormone, stimulus_strength: f64) {
        let max_rate = hormone.max_secretion_rate();
        let amount = (stimulus_strength * max_rate).clamp(0.0, max_rate);

        if let Some(level) = self.levels.get_mut(&hormone) {
            level.secrete(amount);
        }
    }

    /// Get privilege level for an action (receptor-mediated response)
    /// 
    /// PATENT CLAIM: Privilege as receptor density and response curve
    pub fn privilege(&self, hormone: Hormone) -> f64 {
        let level = self.levels.get(&hormone).map(|l| l.level).unwrap_or(0.5);
        let receptor = self.receptors.get(&hormone);

        match receptor {
            Some(r) => r.response(level),
            None => level, // Fallback to raw level
        }
    }

    /// Compute overall alignment (homeostasis indicator)
    /// 
    /// High alignment = all hormones near baseline = stable system
    pub fn alignment(&self) -> f64 {
        let baseline = 0.5;
        let mut deviation_sum = 0.0;
        let mut count = 0.0;

        for level in self.levels.values() {
            deviation_sum += (level.level - baseline).abs();
            count += 1.0;
        }

        if count > 0.0 {
            1.0 - (deviation_sum / count) // Higher when closer to baseline
        } else {
            1.0
        }
    }

    /// Get the dominant hormone (highest above baseline)
    pub fn dominant_hormone(&self) -> Option<Hormone> {
        let baseline = 0.5;
        self.levels
            .iter()
            .filter(|(_, l)| l.level > baseline)
            .max_by(|(_, a), (_, b)| {
                a.level.partial_cmp(&b.level).unwrap_or(std::cmp::Ordering::Equal)
            })
            .map(|(h, _)| *h)
    }

    /// Apply negative feedback: high hormone levels inhibit further secretion
    /// 
    /// PATENT CLAIM: Negative feedback loop for self-regulation
    pub fn apply_negative_feedback(&mut self, hormone: Hormone) -> f64 {
        let level = self.levels.get(&hormone).map(|l| l.level).unwrap_or(0.5);

        // Feedback factor: reduces secretion at high levels
        // At level 0.5 → factor 1.0 (no feedback)
        // At level 0.9 → factor 0.2 (strong inhibition)
        if level > 0.5 {
            1.0 - 0.8 * ((level - 0.5) / 0.5).powi(2)
        } else {
            1.0
        }
    }

    /// Convert to 8-dimensional reputation vector
    pub fn to_reputation_vector(&self) -> [f64; 8] {
        [
            self.levels.get(&Hormone::Cortisol).map(|l| l.level).unwrap_or(0.5),
            self.levels.get(&Hormone::Oxytocin).map(|l| l.level).unwrap_or(0.5),
            self.levels.get(&Hormone::Serotonin).map(|l| l.level).unwrap_or(0.5),
            self.levels.get(&Hormone::Dopamine).map(|l| l.level).unwrap_or(0.5),
            self.levels.get(&Hormone::Adrenaline).map(|l| l.level).unwrap_or(0.5),
            self.levels.get(&Hormone::Endorphins).map(|l| l.level).unwrap_or(0.5),
            self.levels.get(&Hormone::Norepinephrine).map(|l| l.level).unwrap_or(0.5),
            self.levels.get(&Hormone::GrowthHormone).map(|l| l.level).unwrap_or(0.5),
        ]
    }

    /// Create from 8-dimensional reputation vector
    pub fn from_reputation_vector(agent_id: [u8; 32], vector: [f64; 8]) -> Self {
        let mut state = Self::new(agent_id);

        let hormones = Hormone::all();
        for (i, hormone) in hormones.iter().enumerate() {
            if let Some(level) = state.levels.get_mut(hormone) {
                level.level = vector[i].clamp(0.0, 1.0);
                level.peak = level.level;
            }
        }

        state
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_half_life_decay() {
        let mut level = HormoneLevel::new(0.9);
        let half_life = Hormone::Dopamine.half_life(); // 5 min

        // After one half-life, should be halfway to baseline
        level.decay(half_life, half_life);

        // Started at 0.9, baseline 0.5, so 0.5 + (0.9-0.5)*0.5 = 0.7
        assert!((level.level - 0.7).abs() < 0.01);
    }

    #[test]
    fn test_receptor_saturation() {
        let receptor = HormoneReceptor::new(Hormone::Oxytocin);

        // At low levels, response is nearly linear
        let low_response = receptor.response(0.1);

        // At high levels, response saturates
        let high_response = receptor.response(0.9);

        // Saturation: diminishing returns
        assert!(high_response / low_response < 9.0);
    }

    #[test]
    fn test_negative_feedback() {
        let mut state = EndocrineState::new([0u8; 32]);

        // Secrete dopamine
        state.secrete(Hormone::Dopamine, 0.8);
        let fb1 = state.apply_negative_feedback(Hormone::Dopamine);

        // Higher level = stronger feedback inhibition
        state.secrete(Hormone::Dopamine, 0.5);
        let fb2 = state.apply_negative_feedback(Hormone::Dopamine);

        assert!(fb2 < fb1, "Higher levels should have stronger feedback");
    }

    #[test]
    fn test_alignment_at_baseline() {
        let state = EndocrineState::new([0u8; 32]);

        // All at baseline = perfect alignment
        let alignment = state.alignment();
        assert_eq!(alignment, 1.0);
    }

    #[test]
    fn test_circadian_rhythm() {
        let level = HormoneLevel {
            level: 0.5,
            peak: 0.5,
            last_updated: 0.0,
            circadian_phase: 0.0,
        };

        // At 6 AM (21600) vs 6 PM (64800) - opposite phases of the day
        let morning = level.effective_level(21600.0);  // 6 AM - rising phase
        let evening = level.effective_level(64800.0);  // 6 PM - falling phase

        // There should be variation across the day
        assert!(
            (morning - evening).abs() > 0.01 || morning != evening,
            "Circadian should vary: morning={}, evening={}",
            morning,
            evening
        );
    }
}
